### ISSUE TITLE
Incorrect ONNX Export for Unidirectional Broadcasting in PReLU

### ISSUE THREAD

**Issue by ganler** [pytorch/pytorch#70570]
### 🐛 Describe the bug

```python
import torch
from tvm import relay
import onnx

model = torch.nn.PReLU()

onnx_model = "test.onnx"

torch.onnx.export(model, (torch.tensor(1.0),), onnx_model, verbose=True)

mod, params = relay.frontend.from_onnx(onnx.load(onnx_model))

mod = relay.transform.InferType()(mod)
```

When compiling this **PReLU** operator in ONNX, TVM will fail due to that the given model is incorrect. After some investigation, I found that in ONNX's [specification](https://github.com/onnx/onnx/blob/master/docs/Operators.md#prelu):

> This operator supports unidirectional broadcasting (tensor slope should be **unidirectional broadcastable** to input tensor X);

This means in a valid ONNX model, rank of `slope` must be <= rank of PReLU's input (Unidirectional Broadcasting) while PyTorch's PReLU seems to allow 1D `slope` when input is a scalar. So to make everything compatible, during ONNX exportation, PyTorch end should add an `unsqueeze` operation for the input.

By editing `symbolic_opeset9.py`:

```python
def prelu(g, self, weight):
    self_rank = sym_help._get_tensor_rank(self)
    if self_rank is not None and self_rank > 2:
        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+       return g.op("PRelu", sym_help._unsqueeze_helper(g, self, [0]), weight)
    return g.op("PRelu", self, weight)
```

can fix the issue. :-)

### Versions

PyTorch version: 1.10.0+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 11.1.0
Clang version: 13.0.0
CMake version: version 3.22.1
Libc version: glibc-2.33

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.10.84-1-MANJARO-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.5.119
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Ti
Nvidia driver version: 495.44
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.3.0
/usr/lib/libcudnn_adv_infer.so.8.3.0
/usr/lib/libcudnn_adv_train.so.8.3.0
/usr/lib/libcudnn_cnn_infer.so.8.3.0
/usr/lib/libcudnn_cnn_train.so.8.3.0
/usr/lib/libcudnn_ops_infer.so.8.3.0
/usr/lib/libcudnn_ops_train.so.8.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.0rc3
[pip3] numpydoc==1.1.0
[pip3] torch==1.10.0+cu113
[pip3] torch-tb-profiler==0.3.1
[pip3] torchaudio==0.10.0+cu113
[pip3] torchvision==0.11.1+cu113
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py38h27cfd23_1  
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.1            py38ha9443f7_2  
[conda] mypy_extensions           0.4.3                    py38_0  
[conda] numpy                     1.22.0rc3                pypi_0    pypi
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] torch                     1.10.0+cu113             pypi_0    pypi
[conda] torch-tb-profiler         0.3.1                    pypi_0    pypi
[conda] torchaudio                0.10.0+cu113             pypi_0    pypi
[conda] torchvision               0.11.1+cu113             pypi_0    pypi

**Comment by garymm** [comment]
From [the spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#inputs-77):

> The shape of slope can be smaller then first input X; if so, its shape must be unidirectional broadcastable to X

So I don't think the spec requires `rank(slope) <= rank(X)`.

**Comment by ganler** [comment]
@garymm Thanks for pointing it out. You are right! Basically I think `rank(slope) <= rank(X)` is not necessary as long as the rank of `slope` after "rank minimization" <= `rank(X)`. But it seems making `slope`'s rank as low as possible will make things more compatible. :-)

**Comment by ganler** [comment]
Also, ONNXRuntime won't throw an error but it will raise a warning:

> "Expected shape from model of {} does not match actual shape of {1} for output 2"

**Comment by garymm** [comment]
I don't get any warnings printed for this code:

```python
import onnxruntime
import torch

model = torch.nn.PReLU()
onnx_model = "test.onnx"
torch.onnx.export(model, (torch.tensor(1.0),), onnx_model, verbose=True)
i_sess = onnxruntime.InferenceSession(onnx_model)
i_sess.run([], {"input": torch.tensor(1.0).numpy(),})
```

What code are you using to produce the error, and are you sure it's related to this issue?

From my current understanding, this is a bug in TVM Relay, not torch.onnx.

**Comment by ganler** [comment]
Hi @garymm , maybe it is because the ONNX version.

```python
import onnxruntime
import torch

model = torch.nn.PReLU()
onnx_model = "test.onnx"
torch.onnx.export(model, (torch.tensor(1.0),), onnx_model, verbose=True)
i_sess = onnxruntime.InferenceSession(onnx_model)
i_sess.run([], {"input": torch.tensor(1.0).numpy(),})

print(f'{onnxruntime.__version__=}; {torch.__version__=}')
```

Output:
```
graph(%input : Float(requires_grad=0, device=cpu),
      %weight : Float(1, strides=[1], requires_grad=1, device=cpu)):
  %2 : Float(requires_grad=1, device=cpu) = onnx::PRelu(%input, %weight) # /home/ganler/anaconda3/lib/python3.8/site-packages/torch/nn/functional.py:1500:0
  return (%2)

2022-02-01 19:33:40.623790449 [W:onnxruntime:, execution_frame.cc:811 VerifyOutputSizes] Expected shape from model of {} does not match actual shape of {1} for output 2
onnxruntime.__version__='1.10.0'; torch.__version__='1.10.2+cu102'
```

Thanks for your help, I think TVM community can optimize this as well.

**Comment by ganler** [comment]
Maybe you are right. I tried latest TensorRT and it is fine!

**Comment by garymm** [comment]
I'll have to dig into this a bit more to be sure I understand the source of the error message and see if your PR is the right fix.
If you've already done that digging feel free to explain. Otherwise it may be a few days until I have time.

**Comment by garymm** [comment]
I dug in. I believe that there are a few issues:
1. The constraint `rank(slope) <= rank(X)` makes sense, but the ONNX spec doesn't explicitly say that. That should be fixed in the spec and the ONNX checker. I [reported this](https://github.com/onnx/onnx/issues/3651#issuecomment-1034378475).
2. In general ONNX Runtime (and it seems TensorRT in this case) is permissive about using a scalar interchangeably with a tensor of 1 element. So this is a special case violation to the above constraint that arguably makes sense, but it's not surprising that not all ONNX implementations support it.

In conclusion I agree this should be improved in the exporter to guarantee `rank(slope) <= rank(X)` and I will review the PR.
